src/outpack/location.py

```diff
@@ -71,6 +82,149 @@ def outpack_location_rename(old, new, root=None, *, locate=True):
+        location.mark_known()
+
+    # TODO: deorphan recovered packets
+
+
+class OutpackLocation:
+    def __init__(self, location_name, root):
+        self.__root = root
+        self.location_name = location_name
+        self.__driver = _location_driver(location_name, root)
+        self._hint_remove = (
+            f'Probably all you can do at this point is '
+            f'remove this location from your configuration '
+            f'by running '
+            f'orderly_location_remove("{self.location_name}")'
+        )
+
+    def pull_all_metadata(self):
+        known_there = self.__driver.list()
+        known_here = self.__root.index.data().metadata.keys()
+        for packet_id in known_there:
+            if packet_id not in known_here:
+                self.pull_packet_metadata(packet_id)
+
+    def pull_packet_metadata(self, packet_id):
+        metadata = self.__driver.metadata(packet_id)[packet_id]
+        expected_hash = self.__driver.list()[packet_id].hash
+
+        hash_validate_string(
+            metadata,
+            expected_hash,
+            f"metadata for '{packet_id}' from '{self.location_name}'",
+            [
+                "This is bad news, I'm afraid. Your location is sending data "
+                "that does not match the hash it says it does. Please let us "
+                "know how this might have happened.",
+                self._hint_remove,
+            ],
+        )
+
+        path_metadata = self.__root.path / ".outpack" / "metadata"
+        os.makedirs(path_metadata, exist_ok=True)
+        filename = path_metadata / packet_id
+        with open(filename, "w") as f:
+            f.writelines(metadata)
+
+    def validate_hashes(self, packets: List[PacketLocation]):
+        mismatched_hashes = set()
+        known_there = self.__driver.list()
+        for packet in packets:
+            if known_there.get(packet.packet) is not None:
+                hash_there = known_there[packet.packet].hash
+                hash_here = packet.hash
+                if hash_there != hash_here:
+                    mismatched_hashes.add(packet.packet)
+
+        if len(mismatched_hashes) > 0:
+            id_text = "', '".join(mismatched_hashes)
+            msg = (
+                f"Location '{self.location_name}' has conflicting metadata\n"
+                f"This is really bad news. We have been offered metadata "
+                f"from '{self.location_name}' that has a different hash to "
+                f"metadata that we have already imported from other "
+                f"locations. I'm not going to import this new metadata, "
+                f"but there's no guarantee that the older metadata is "
+                f"actually what you want!\nConflicts for: '{id_text}'\n"
+                f"We would be interested in this case, please let us know\n"
+                f"{self._hint_remove}"
+            )
+            raise Exception(msg)
+
+    def mark_known(self):
```

Comment:
I am confused about this one - the equivalent I can see in the R version is "mark this packet known", rather than all packets

Reply:
Am I reading this wrong? From https://github.com/mrc-ide/orderly2/blob/main/R/location.R#L477 to 482

This:
1. Get's the list of packets from our `.outpack/location/<location_name>` (i.e. the list of packets we know about from the location we are pulling from)
2. Get's the list of packets known at the remote path which are not known about here, so any new packets from the remote
3. Marks those as known (i.e. saves metadata to `.outpack/location/<location_name>`

That's the same as the python impl here?

Reply:
Oh maybe just naming confusion? I've refactored tided this up a bit based on other comments earlier, see `_mark_all_known` now, should be the equivalent of block of R linked in previous comment